### DISCUSSION TITLE
v4.0.0-alpha.1

### DISCUSSION THREAD

**Discussion by mcollina** [fastify/fastify#3710] [General]
#  Fastify v4.0.0-alpha.1

This is first release of Fastify v4. We will soon be issuing release candidates.

## What's Changed

* Checked if error handler is default before setting headers by @Swampr in https://github.com/fastify/fastify/pull/2609
* feat: set undefined on null input by @MetCoder95 in https://github.com/fastify/fastify/pull/2731
* chore: merge master with next  by @salmanm in https://github.com/fastify/fastify/pull/2753
* Set exposeHeadRoutes: true by default by @mcollina in https://github.com/fastify/fastify/pull/2826
* dropped flatstrt by @mcollina in https://github.com/fastify/fastify/pull/3016
* Typed decorators by @wyozi in https://github.com/fastify/fastify/pull/2981
* feat: Improve error experience by @MetCoder95 in https://github.com/fastify/fastify/pull/2954
* Allow async handler to resolve with undefined by @sergejostir in https://github.com/fastify/fastify/pull/2702
* Remove overdue deprecations by @sergejostir in https://github.com/fastify/fastify/pull/3070
* Remove content type coercion for response streams by @luke88jones in https://github.com/fastify/fastify/pull/3086
* feat: Handle new `avvio` error codes by @MetCoder95 in https://github.com/fastify/fastify/pull/3106
* fix: handle invalid url by @climba03003 in https://github.com/fastify/fastify/pull/3128
* Clarification for "Simplify reply sent monitoring (#3072)" by @sergejostir in https://github.com/fastify/fastify/pull/3132
* Deprecate modifying reply.sent property by @sergejostir in https://github.com/fastify/fastify/pull/3140
* Fix responding from a hook doc by @sergejostir in https://github.com/fastify/fastify/pull/3161
* Deem everything thrown as an error by @sergejostir in https://github.com/fastify/fastify/pull/3200
* Encapsulated error handling by @mcollina in https://github.com/fastify/fastify/pull/3261
* Payload validation by @matthyk in https://github.com/fastify/fastify/pull/3274
* Ajv8 by default by @Eomm in https://github.com/fastify/fastify/pull/3271
* Update types for Pino 7 by @kibertoad in https://github.com/fastify/fastify/pull/3281
* Update dependencies by @kibertoad in https://github.com/fastify/fastify/pull/3294
* chore(deps): Update to new Avvio major version by @MetCoder95 in https://github.com/fastify/fastify/pull/3288
* Update to pino rc4 by @kibertoad in https://github.com/fastify/fastify/pull/3302
* Increase http2SessionTimeout and keepAliveTimeout to 72 seconds by @mcollina in https://github.com/fastify/fastify/pull/3304
* fix(typescript): allow to pass all pino options by @climba03003 in https://github.com/fastify/fastify/pull/3383
* Fastify Type Providers by @sinclairzx81 in https://github.com/fastify/fastify/pull/3398
* docs: Extend TypeBox documentation by @Jnig in https://github.com/fastify/fastify/pull/3437
* Adjust existing PRs for the new reply.sent handing by @sergejostir in https://github.com/fastify/fastify/pull/3444
* Skip encapsulation of Request and Reply when not needed by @mcollina in https://github.com/fastify/fastify/pull/3479
* fix: Refactor Context constructor and avoid positional arguments by @vtcaregorodtcev in https://github.com/fastify/fastify/pull/3483
* remove app.use and fix middleware tests by @genzyy in https://github.com/fastify/fastify/pull/3506
* chore: bump find-my-way to 5.1.0 by @anonrig in https://github.com/fastify/fastify/pull/3515
* Type Provider Reply Types by @sinclairzx81 in https://github.com/fastify/fastify/pull/3524
* chore: bump dependencies by @anonrig in https://github.com/fastify/fastify/pull/3531
* build: drop node 12 support by @anonrig in https://github.com/fastify/fastify/pull/3532
* Remove .writableEnded fallbacks for node12 by @sergejostir in https://github.com/fastify/fastify/pull/3590
* upgrade fast-json-stringify with ajv8 by @Eomm in https://github.com/fastify/fastify/pull/3280
* fix: bad headers characters by @Eomm in https://github.com/fastify/fastify/pull/3593
* feat: ajv default array coercion by @Eomm in https://github.com/fastify/fastify/pull/3594
* docs: add type provider encapsulation by @RafaelGSS in https://github.com/fastify/fastify/pull/3647
* localhost binds to ipv4 and ipv6 by @Eomm in https://github.com/fastify/fastify/pull/3606
* feat: bump ajv-compiler by @zekth in https://github.com/fastify/fastify/pull/3687
* Tests should pass on a host with only a single network interface by @mcollina in https://github.com/fastify/fastify/pull/3703
* fix: call normal 404 handler on unsupported method by @markwainwright in https://github.com/fastify/fastify/pull/3705

## New Contributors
* @Swampr made their first contribution in https://github.com/fastify/fastify/pull/2609
* @luke88jones made their first contribution in https://github.com/fastify/fastify/pull/3086
* @sinclairzx81 made their first contribution in https://github.com/fastify/fastify/pull/3398
* @Jnig made their first contribution in https://github.com/fastify/fastify/pull/3437
* @vtcaregorodtcev made their first contribution in https://github.com/fastify/fastify/pull/3483
* @anonrig made their first contribution in https://github.com/fastify/fastify/pull/3515
* @markwainwright made their first contribution in https://github.com/fastify/fastify/pull/3705

**Full Changelog**: https://github.com/fastify/fastify/compare/v3.27.2...v4.0.0-alpha.1

<hr /><em>This discussion was created from the release <a href='https://github.com/fastify/fastify/releases/tag/v4.0.0-alpha.1'>v4.0.0-alpha.1</a>.</em>

**Comment by gajus** [comment]
Nothing mentioned in the log (appears) to be a breaking change. What's the reason for v4?

**Reply by mcollina** [reply]
Most of the changes above are actually breaking changes.

**Reply by jsumners** [reply]
There's plenty of breaking changes. Particularly around async handlers, error handling, and supported Node versions.

**Reply by gajus** [reply]
Generally, it would be helpful to tag each breaking change with "[breaking change]".

It was not immediately obvious reading these release notes which are breaking changes.

**Reply by Eomm** [reply]
We will publish a detailed changelog when the `alpha` stage is complete.

Maybe is it worth publishing a draft?
